src/backend/distributed/commands/table.c

```diff
@@ -593,59 +616,213 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
+	bool leftIsNonDistTable = IsCitusTableType(leftRelationId,
+											   CITUS_TABLE_WITH_NO_DIST_KEY);
+	if (leftIsNonDistTable)
 	{
-		Oid rightRelationId = RangeVarGetRelid(constraint->pktable, lockmode,
-											   alterTableStatement->missing_ok);
-		if (IsCitusTableType(rightRelationId, CITUS_LOCAL_TABLE))
+		return rightHasPostgresLocalTable;
+	}
+
+	return false;
+}
+
+
+/*
+ * ConvertPostgresLocalTablesToCitusLocalTables converts each postgres table
+ * involved in foreign keys to be defined by given alter table command and the
+ * other tables connected to them via a foreign key graph to citus local tables.
+ */
+static void
+ConvertPostgresLocalTablesToCitusLocalTables(AlterTableStmt *alterTableStatement)
+{
+	List *rightRelationRangeVarList =
+		GetAlterTableAddFKeyRightRelationRangeVarList(alterTableStatement);
+	RangeVar *leftRelationRangeVar = alterTableStatement->relation;
+	List *relationRangeVarList = lappend(rightRelationRangeVarList, leftRelationRangeVar);
+
+	/*
+	 * To prevent deadlocks, sort the list before converting each postgres local
+	 * table to a citus local table.
+	 */
+	relationRangeVarList = SortList(relationRangeVarList, CompareRangeVarsByOid);
+
+	/*
+	 * Here we should operate on RangeVar objects since relations oid's would
+	 * change in below loop due to CreateCitusLocalTable.
+	 */
+	RangeVar *relationRangeVar;
+	foreach_ptr(relationRangeVar, relationRangeVarList)
+	{
+		List *commandList = alterTableStatement->cmds;
+		LOCKMODE lockMode = AlterTableGetLockLevel(commandList);
+		bool missingOk = alterTableStatement->missing_ok;
+		Oid rightRelationId = RangeVarGetRelid(relationRangeVar, lockMode, missingOk);
+		if (!OidIsValid(rightRelationId))
 		{
-			ErrorOutForFKeyBetweenPostgresAndCitusLocalTable(leftRelationId);
+			/*
+			 * As we are in preprocess, missingOk might be true and relation
+			 * might not exist.
+			 */
+			continue;
 		}
+		else if (IsCitusTable(rightRelationId))
+		{
+			/*
+			 * relationRangeVarList has also reference and citus local tables
+			 * involved in this ADD FOREIGN KEY command. Morever, even if
+			 * rightRelationId was belonging to a postgres  local table initially,
+			 * we might had already converted it to a citus local table by cascading.
+			 */
+			continue;
+		}
+
+		/*
+		 * CopyErrorData() requires (CurrentMemoryContext != ErrorContext),
+		 * so we store CurrentMemoryContext here.
+		 */
+		MemoryContext savedMemoryContext = CurrentMemoryContext;
+		PG_TRY();
+		{
+			bool cascade = true;
+			CreateCitusLocalTable(rightRelationId, cascade);
+		}
+		PG_CATCH();
+		{
+			MemoryContextSwitchTo(savedMemoryContext);
+
+			ErrorData *errorData = CopyErrorData();
+			FlushErrorState();
+
+			if (errorData->elevel != ERROR)
+			{
+				/* don't try to intercept PANIC or FATAL */
+				PG_RE_THROW();
+			}
+
+			/* extend error message with foreign key creation error */
+			StringInfo errorMessage = makeStringInfo();
```

Comment:
Do we really need this try/catch block? We are conveying the same error message to the user with a  `cannot create foreign key` prefix, which doesn't sound crucial?